### PR TITLE
virtio-net: add support for notification suppression

### PR DESCRIPTION
**Pull request by bchalios**
# Reason for This PR

Integrate notification suppression for the VirtIO net device (see issues: #2478, #2636) using functionality added for the 
equivalent block device feature (#2658).

## Description of Changes

Here I mainly used the machinery introduced by @serban300 for the corresponding block device feature. I ran a number of tests to evaluate performance impact.

My experimental setup is the following:
* I 'm using an `c5d.metal` instance
* I am using 48 cores to avoid using hyper-threading
* Each microVM has 5 vcpus and 4GB of memory (so that each microVM has 7 threads in total)
* I launch from 1-12 microVMs pinning them in cores and forcing them to allocate memory from the corresponding NUMA node, i.e. 6 microVMs fill up the available cores. Data points with more than 6 microVMs represent overcommit scenarios.
* Each microVM is running an `iperf3` server
* On the host side I launch an `iperf3` client (which I pin in a core not occupied by a microVM thread) per microVM and launch one TX and one RX experiment.

In the following table I show the percentage increase of bandwidth (RX and TX) when using notification over the current main.

| #VMs | RX (% increase) | TX (% increase) |
|---|---|---|
|1|98.53|98.53|
|2|98.48|98.48|
|3|97.64|97.64|
|4|100.59|100.59|
|5|100.78|100.78|
|6|99.04|99.04|
|7|97.83|97.83|
|8|98.92|98.92|
|9|99.52|99.52|
|10|98.64|98.64|
|11|99.83|99.83|
|12|100.00|100.00|

We see that the current implementation does not actually provide any performance boost. My current best guess is that
due to the fact that our virtio-net implementation already consumes as many Descriptor heads from the virtqueue as possible before raising a trigger, suppressing notification does not really make a difference.

Trying to dive a bit deeper, I also compared the `net` metrics of the two implementations:

| metric | notification_suppression / main (%) |
|---|---|
|tx_packets_count|100.71|
|no_rx_avail_buffer|112.94|
|rx_queue_event_count|101.77|
|rx_packets_count|102.58|
|rx_tap_event_count|117.18|
|tx_queue_event_count|100.65|
|no_tx_avail_buffer|0.45|
|tx_count|100.71|
|tx_bytes_count|100.71|
|rx_bytes_count|102.59|
|rx_count|102.58|

An interesting result from this table is that whereas notification suppression almost eliminates `no_tx_avail_buffer` events, the corresponding RX path events (`no_rx_avail_buffer`) actually increase by ~13%.

- [ ] This functionality can be added in [`rust-vmm`][1].

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The issue which led to this PR has a clear conclusion.
- [ ] This PR follows the solution outlined in the related issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes follow the [Runbook for Firecracker API changes][2].
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

File: src/devices/src/virtio/net/device.rs
Comment (serban300) [inline src/devices/src/virtio/net/device.rs:338]: I don't know if this guarantees that we won't receive any error. In the end GuestMemory will issue a `write` syscall. Are we sure that it will complete successfully ? Also are we sure we're not going to receive other errors that are not related to `len` ? It would be risky to do this.

File: src/devices/src/virtio/net/device.rs
Comment (serban300) [inline src/devices/src/virtio/net/device.rs:262]: Can we delete `rx_deferred_irqs` and rely only on `prepare_kick()` ? It's a tradeoff for code simplicity. It will degrade the performance when the guest doesn't use notification suppression but usually this shouldn't happen. I think we did something similar for the block device.

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs:338]: I left this a TODO because I wanted to examine it better, however reading from the `write_slice` documentation:

```
Returns an error if ther isn't enuogh space within the container to write the entire slice.
Part of the data may have been copied nevertheless.
```

**Comment by bchalios** [comment]
For completeness, I also ran experiments to compare CPU load with notification suppression and without.


RX without notification suppression:

| #VMs | Bandwidth | VMM thread CPU load | kworker CPU load | Bandwidth per CPU load |
|------|-----------|---------------------|------------------|------------------------|
|1|40.8|0.81|0.03|48.55|
|2|79.6|1.63|0.05|47.34|
|3|111.9|2.46|0.07|44.24|
|4|152.7|3.28|0.10|45.23|
|5|192.3|4.10|0.12|45.62|
|6|228.5|4.90|0.14|45.33|
|7|242.9|5.43|0.16|43.47|
|8|258.3|5.98|0.18|41.96|
|9|270.9|6.39|0.19|41.18|
|10|284.7|6.93|0.21|39.89|
|11|297.0|7.39|0.23|39.00|
|12|306.7|7.77|0.23|38.33|

RX with notification suppression:

| #VMs | Bandwidth | VMM thread CPU load | kworker CPU load | Bandwidth per CPU load |
|------|-----------|---------------------|------------------|------------------------|
|1|40.7|0.82|0.02|48.35|
|2|79.5|1.64|0.05|46.96|
|3|113.0|2.47|0.07|44.46|
|4|154.2|3.28|0.09|45.69|
|5|194.2|4.09|0.11|46.19|
|6|228.8|4.93|0.14|45.11|
|7|244.7|5.45|0.17|43.53|
|8|258.5|5.98|0.17|42.04|
|9|269.0|6.42|0.19|40.70|
|10|284.4|6.97|0.22|39.59|
|11|298.0|7.50|0.23|38.57|
|12|308.9|7.98|0.25|37.50|

TX without notification suppression:

| #VMs | Bandwidth | VMM thread CPU load | kworker CPU load | Bandwidth per CPU load |
|------|-----------|---------------------|------------------|------------------------|
|1|31.3|0.81|0.02|37.74|
|2|59.8|1.63|0.04|35.87|
|3|87.7|2.44|0.05|35.21|
|4|117.5|3.24|0.08|35.49|
|5|145.6|4.05|0.10|35.06|
|6|172.5|4.90|0.11|34.42|
|7|196.6|5.52|0.12|34.80|
|8|211.9|6.09|0.13|34.03|
|9|223.1|6.66|0.15|32.76|
|10|248.1|7.27|0.16|33.36|
|11|262.1|7.85|0.19|32.62|
|12|273.6|8.30|0.17|32.29|

TX with notification suppression:

| #VMs | Bandwidth | VMM thread CPU load | kworker CPU load | Bandwidth per CPU load |
|------|-----------|---------------------|------------------|------------------------|
|1|31.6|0.81|0.02|37.93|
|2|61.1|1.63|0.04|36.53|
|3|88.7|2.46|0.06|35.22|
|4|118.9|3.29|0.08|35.30|
|5|145.4|4.09|0.09|34.72|
|6|169.2|4.89|0.11|33.87|
|7|196.6|5.54|0.14|34.61|
|8|212.4|6.07|0.16|34.06|
|9|224.2|6.62|0.14|33.15|
|10|247.1|7.25|0.15|33.40|
|11|263.9|7.89|0.18|32.72|
|12|276.3|8.33|0.17|32.52|

It seems that there isn't much of a difference regarding CPU load either.

File: src/devices/src/virtio/net/device.rs
Comment (serban300) [inline src/devices/src/virtio/net/device.rs:271]: Can we have a single method that receives the queue as a parameter ? Probably as an enum.

File: src/devices/src/virtio/net/device.rs
Comment (serban300) [inline src/devices/src/virtio/net/device.rs:686]: Nit: I wouldn't have a separate commit for this. I would merge it in the 2nd or 3rd commit. Wherever it fits best.

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs]: I got rid of the `rx_deferred_irqs` altogether. I 'm now relying only on `prepare_kick()` to trigger interrupts

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs]: I added a commit that does that.

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs:678]: done!

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs:325]: Agreed, that makes sense!

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs]: done!

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs]: done

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs]: that makes sense. done!

**Review by dianpopa — COMMENTED** [review]
The PR looks good to me.
I guess there are a couple of things left to do:
* measure latency with prepare_kick as the main mechanism
* based on the performance results, have a discussion on the actual benefits and next steps

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs]: done!

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs]: done!

**Review by serban300 — COMMENTED** [review]
Nice job ! Looks good. On my side there are 2 comments left, and after that I think we are good to merge the PR.

File: src/devices/src/virtio/net/device.rs
Comment (bchalios) [reply src/devices/src/virtio/net/device.rs:1583]: done

File: tests/framework/artifacts.py
Comment (bchalios) [reply tests/framework/artifacts.py:307]: I 'll wait until uffd is merged before doing that

**Comment by bchalios** [comment]
So, there are some new numbers comparing against the latest main (uffd merged). I added latency measurements as well  using ping (from the host to the microVM).

The tables compare the two cases and the numbers present % difference. `(baseline - profiled) / baseline`, i.e. throughput positive is better, for latency negative is better.

RX:

| #VMs | Bandwidth | VMM thread CPU load | kworker CPU load | Bandwidth per CPU load |
|------|-----------|---------------------|------------------|------------------------|
|1|2.23|1.16|0.00|0.64
|2|-1.98|0.00|0.00|-2.20
|3|-2.29|0.00|-11.11|-2.32
|4|1.03|0.29|9.09|0.57
|5|-0.15|0.23|-6.67|-0.20
|6|0.94|-0.19|0.00|0.96
|7|2.28|1.75|0.00|0.50
|8|-1.32|-2.69|0.00|1.28
|9|-0.43|-0.29|0.00|-0.03
|10|-0.75|0.55|0.00|-1.35
|11|-2.03|-0.13|0.00|-1.78
|12|-2.28|-2.26|3.23|-0.24

TX:

| #VMs | Bandwidth | VMM thread CPU load | kworker CPU load | Bandwidth per CPU load |
|------|-----------|---------------------|------------------|------------------------|
|1|1.00|2.41|0.00|-0.77
|2|1.52|1.21|25.00|0.17
|3|2.12|0.80|16.67|0.54
|4|0.86|0.90|0.00|-0.03
|5|1.80|0.96|0.00|0.86
|6|2.58|0.40|15.38|1.86
|7|-0.15|-0.34|6.25|0.00
|8|0.82|0.93|5.56|-0.24
|9|-0.91|-0.99|-5.26|0.09
|10|1.81|0.52|-4.76|1.28
|11|1.02|0.24|0.00|0.69
|12|0.52|0.11|-4.35|0.45

Ping :

| #VMs | min | avg | max | mdev |
|------|-----|-----|-----|------|
|1|-61.38|-12.45|-3.11|55.77
|2|16.53|1.80|-2.35|-29.82
|3|-0.70|-5.59|-0.85|-18.87
|4|-26.72|-9.97|-1.72|35.29
|5|13.51|-2.41|-71.31|-74.71
|6|53.93|6.07|0.29|-19.23
|7|-53.50|-9.79|-3.94|95.24
|8|1.85|-2.71|-4.41|-11.11
|9|16.03|5.59|11.30|-10.20
|10|-15.18|-0.35|-6.43|-3.28
|11|-2.84|-0.66|-5.99|-6.00
|12|-42.95|0.68|-4.28|16.67

I think that performance-wise there is not much difference after enabling notification suppression. So, the decision we have to take is whether the patch essentially simplifies the implementation and is desirable for symmetry of implementation between block and net devices.

**Review by alsrdn — APPROVED** [review]

**Review by dianpopa — APPROVED** [review]

**Pull request merged**
